modules/govuk/files/usr/local/bin/govuk_supervised_initctl

```diff
@@ -0,0 +1,137 @@
+#!/bin/bash
+
+# This script is a wrapper around upstart initctl start/restart/reload
+# commands and is intended for applications/service that have their main
+# process in a pid file at /var/run/$sevice/app.pid - which is the case for
+# most GOV.UK applications
+#
+# The purpose of this script is to monitor whether the running of these
+# commands results in the process actually changing - whether it be started,
+# restarted or reloaded - as this is not something running initctl has
+# sufficient visibility for (for example you can run `initctl start $app`
+# and find upstart is satisfied the app started when it crashed on boot)
+#
+# This script is pretty heavily nuanced towards GOV.UK's use cases and isn't
+# intended to be a generic initctl wrapper for all use cases. I'm also
+# somewhat convinced there are better utilities to manage this problem in
+# newer ubuntu releases.
+
+if [ "$#" -ne "2" ]; then
+  echo "Usage: sudo govuk_supervised_initctl <COMMAND> <SERVICE>"
+  exit 1
+fi
+
+set -eu
+
+status () {
+  echo "---> ${@}"
+}
+
+error () {
+  echo "ERROR: ${@}" >&2
+  exit 1
+}
+
+pid_running () {
+  ps -p ${@} &> /dev/null
+}
+
+command=$1
+service=$2
+pid_file=/var/run/$service/app.pid
+original_pid="" && [ -e $pid_file ] && original_pid=$(cat $pid_file)
+
+case $command in
+  start)
+    status "Checking if ${service} is running"
+    if [[ $(initctl status $service) =~ "running" ]]; then
+      if [ ! -z $original_pid ] && pid_running $original_pid; then
+        status "${service} is running"
+        exit 0
+      else
+        status "${service} is an inconsistent state, stopping before starting..."
```

Comment:
Yeah I added this in after some testing when I fell into a situation where it could fail to start/restart without manual intervention. Given, if you have a situation with `sudo govuk_supervised_initctl start #{application} || sudo govuk_supervised_initctl reload #{application}` and the upstart thinks the process is running but the process actually isn't (I found it quite easy to replicate this by putting in a syntax error in a Gemfile). Then you can end up with this code knowing the process isn't running but being upstart not responding to `initctl start` because it think it's running - and the reload command would also fail.

Reply:
That's odd. I assumed that, while it may initially (and correctly) think the service is up (before it fails), it sounds like there's also a lag i.e. the status of according to `initctl` is incorrect. Could we also have another case, then, where `initctl` thinks the process is stopped, but it's actually running?

Reply:
Yeah, I imagine you're right about that second state of running but upstart thinking otherwise. I guess I actually had that covered until I fell into fixing this bug and introduced that scenario. 

I'll have a go at replicating that later on today to see what happens - I'll also get a little more info to explain what happens in the inverse scenario (upstart thinks it's running when it's not).

Reply:
I did some further digging into this. I think the oddness/complexity in this is due to there being two (main) ways of running apps, either directly as per email-alert-service or indirectly through unicornherder.

In the direct scenario things behave as you assumed and the pid file refers to the same pid that upstart is aware of. In the indirect scenario things behave in the annoying way where upstart is happily satisfied that unicornherder is running, whereas the actual app is never started. The pid file for that is the unicorn master instance rather than unicornherder, so whilst upstart is happy the service is running there is no file in /var/run/email-alert-api/app.pid.

To try illustrate with some scenarios:

## Email Alert Service with a typo in the Gemfile

- Service will have a start attempted a few times before quitting
- Upstart won't think the service is running
- There will be a pid file left in /var/run/email-alert-service/app.pid pointing to a non-existent process

## Email Alert API with a typo in the Gemfile

- Upstart thinks the service is running (because unicornherder is)
- There is no pid file created at /var/run/email-alert-api/app.pid
- You can't use `sudo govuk_supervised_initctl restart email-alert-api` because there isn't an old app process running to restart
- You can't run `sudo initctl start email-alert-api` because upstart thinks it's running, therefore needs a `sudo initctl stop email-alert-api` before it can be re-attempted. 

Hope that helps make sense of it

----

You're right that it is possible to have the inverse scenario where a process can be running and Upstart does not have knowledge of it. This scenario seems much less likely to fall into than the previous one (I couldn't think of a problem with an individual app that could cause it for instance), so I've gone with putting that as an error scenario rather than attempting to fix it.

I've got some more documentation to update off the back of this change, but let me know if this all seems sound first.

Reply:
Yes, that does make sense. I checked to see `initctl` refers to the herder PID, yet the `app.pid` file refers to the child / master process, as you described. Conversely, the plain email-alert-service is consistent between the two.

(Apparently initctl does some [auto-cleverness](http://spootnik.org/entries/2014/11/09/pid-tracking-in-modern-init-systems/) to determine the PID a process is running with, which naturally doesn't work with Sidekiq workers because our initctl scripts are actually just helper loops.)

Based on this, I also agree the inverse scenario is unlikely, as it's unrelated to an issue with the application. Let me know when you're done with more documentation and I'll do a final review.

Reply:
Oh cool, that link was an interesting read. It feels like, if you were prepared to go on a big yak shave, you could probably do all the checks of this script through upstart scripts.

A somewhat annoying thing about our Sidekiq upstart scripts is that nothing [actually uses the child process stuff that is part of the loop](https://github.com/alphagov/govuk-puppet/blob/b0bd7e2acbecfc51e12d0c7a547b7f8af03ca844/modules/govuk/templates/procfile-worker.conf.erb#L18-L28) (every app has a worker list of 1) so with a bit of will they could be simplified away.

I've now updated the documentation for this so it should be ready for another review.

Reply:
Thanks, the latest version makes sense to me. I think composition is the right approach here: we only need the extra complexity for the deploy scripts; keeping the individual init scripts simple reduces the amount of code to understand.

After reading through again, I have a couple of minor suggestions as someone trying to understand the purpose of this script in future. One is in a separate comment. The other relates to the consistency between `start` vs `reload` / `restart`: it would be more consistent to check the pidfile first in `start`, like we do in the others.